src/bounces/mod.rs

```diff
@@ -0,0 +1,143 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+  collections::HashMap,
+  error::Error,
+  fmt::{self, Display, Formatter},
+  time::SystemTime,
+};
+
+use rocket::{http::Status, response::Failure};
+
+use auth_db::{BounceRecord, BounceType, Db, DbError};
+use settings::{BounceLimit, BounceLimits, Settings};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Debug)]
+pub struct BounceError
+{
+  pub address: String,
+  pub bounce: Option<BounceRecord>,
+  description: String,
+}
+
+impl BounceError
+{
+  pub fn new(address: &str, bounce: BounceRecord) -> BounceError
+  {
+    let description = format!(
+      "email address violated {} limit",
+      match bounce.bounce_type {
+        BounceType::Hard => "hard bounce",
+        BounceType::Soft => "soft bounce",
+        BounceType::Complaint => "complaint",
+      }
+    );
+
+    BounceError {
+      address: address.to_string(),
+      bounce: Some(bounce),
+      description,
+    }
+  }
+}
+
+impl Error for BounceError
+{
+  fn description(&self) -> &str
+  {
+    &self.description
+  }
+}
+
+impl Display for BounceError
+{
+  fn fmt(&self, formatter: &mut Formatter) -> fmt::Result
+  {
+    write!(formatter, "{}", self.description)
+  }
+}
+
+impl From<DbError> for BounceError
+{
+  fn from(error: DbError) -> BounceError
+  {
+    BounceError {
+      address: String::from(""),
+      bounce: None,
+      description: format!("database error: {}", error.description()),
+    }
+  }
+}
+
+impl From<BounceError> for Failure
+{
+  fn from(_error: BounceError) -> Failure
+  {
+    // Eventually we should be able to do something richer than this,
+    // as per https://github.com/SergioBenitez/Rocket/issues/586.
+    Failure(Status::TooManyRequests)
```

Comment:
This is probably the wrong layer of abstraction for setting the response status, but we're going to properly sort out the error handling in #8 and I figured this will do in the meantime.